File: packages/identity/go/verify.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Verifies a Caracal JWT against an issuer JWKS and enforces zone and scope claims.

package identity

import (
	"errors"
	"fmt"

	"github.com/golang-jwt/jwt/v5"
)

// ErrTokenInvalid signals the token failed JWKS-backed signature or claim validation.
var ErrTokenInvalid = errors.New("token validation failed")

// ErrZoneInvalid signals the zone_id claim is missing or did not match Config.ZoneID.
var ErrZoneInvalid = errors.New("token zone validation failed")

// ErrAgentIdentityRequired signals the token has no agent_session_id.
var ErrAgentIdentityRequired = errors.New("agent identity required")

// ErrDelegationRequired signals the token has no delegation_edge_id.
var ErrDelegationRequired = errors.New("delegation required")

// ErrHopCountExceeded signals the token's hop_count exceeds Config.MaxHopCount.
var ErrHopCountExceeded = errors.New("hop count exceeded")

// ScopeMissingError signals a required scope is absent from the token.
type ScopeMissingError struct {
	Scope string
}

func (e *ScopeMissingError) Error() string {
	return fmt.Sprintf("missing scope: %s", e.Scope)
}

// ChainMismatchError signals a required delegation chain application is absent.
type ChainMismatchError struct {
	ApplicationID string
}

func (e *ChainMismatchError) Error() string {
	return fmt.Sprintf("delegation chain missing application: %s", e.ApplicationID)
}

func readChain(raw any) []ChainHop {
	list, ok := raw.([]any)
	if !ok {
		return nil
	}
	out := make([]ChainHop, 0, len(list))
	for _, item := range list {
		m, ok := item.(map[string]any)
		if !ok {
			continue
		}
		appID, _ := m["app"].(string)
		if appID == "" {
			appID, _ = m["application_id"].(string)
		}
		if appID == "" {
			continue
		}
		session, _ := m["session"].(string)
		if session == "" {
			session, _ = m["agent_session_id"].(string)
		}
		edge, _ := m["edge"].(string)
		if edge == "" {
			edge, _ = m["delegation_edge_id"].(string)
		}
		out = append(out, ChainHop{ApplicationID: appID, AgentSessionID: session, DelegationEdgeID: edge})
	}
	return out
}

func readStringSlice(raw any) []string {
	list, ok := raw.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(list))
	for _, v := range list {
		if s, ok := v.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

// Verify parses and validates a JWT, returning typed Claims on success.
func Verify(tokenStr string, cfg Config) (Claims, error) {
	mapClaims := jwt.MapClaims{}
	_, err := jwt.ParseWithClaims(tokenStr, mapClaims, func(t *jwt.Token) (any, error) {
		kid, _ := t.Header["kid"].(string)
		keys, err := GetJWKS(cfg.Issuer)
		if err != nil {
			return nil, err
		}
		if k, ok := keys[kid]; ok {
			return k, nil
		}
		return nil, jwt.ErrTokenSignatureInvalid
	}, jwt.WithIssuer(cfg.Issuer), jwt.WithAudience(cfg.Audience), jwt.WithValidMethods([]string{jwt.SigningMethodES256.Alg()}))
	if err != nil {
		return Claims{}, ErrTokenInvalid
	}

	scope, _ := mapClaims["scope"].(string)
	zoneID, _ := mapClaims["zone_id"].(string)
	if zoneID == "" || (cfg.ZoneID != "" && zoneID != cfg.ZoneID) {
		return Claims{}, ErrZoneInvalid
	}
	for _, required := range cfg.RequiredScopes {
		if !HasScope(scope, required) {
			return Claims{}, &ScopeMissingError{Scope: required}
		}
	}

	sub, _ := mapClaims["sub"].(string)
	sid, _ := mapClaims["sid"].(string)
	clientID, _ := mapClaims["client_id"].(string)
	agentSessionID, _ := mapClaims["agent_session_id"].(string)
	delegationEdgeID, _ := mapClaims["delegation_edge_id"].(string)
	sourceSessionID, _ := mapClaims["source_session_id"].(string)
	targetSessionID, _ := mapClaims["target_session_id"].(string)
	chain := readChain(mapClaims["delegation_chain"])
	path := readStringSlice(mapClaims["delegation_path"])

	var graphEpoch int64
	switch v := mapClaims["delegation_graph_epoch"].(type) {
	case float64:
		graphEpoch = int64(v)
	case int64:
		graphEpoch = v
	}
	if graphEpoch == 0 {
		switch v := mapClaims["graph_epoch"].(type) {
		case float64:
			graphEpoch = int64(v)
		case int64:
			graphEpoch = v
		}
	}
	var hopCount int
	switch v := mapClaims["hop_count"].(type) {
	case float64:
		hopCount = int(v)
	case int64:
		hopCount = int(v)
	}

	if cfg.RequireAgent && agentSessionID == "" {
		return Claims{}, ErrAgentIdentityRequired
	}
	if cfg.RequireDelegation && delegationEdgeID == "" {
		return Claims{}, ErrDelegationRequired
	}
	if cfg.MaxHopCount > 0 && hopCount > cfg.MaxHopCount {
		return Claims{}, ErrHopCountExceeded
	}
	for _, expected := range cfg.RequireChainContains {
		present := false
		for _, hop := range chain {
			if hop.ApplicationID == expected {
				present = true
				break
			}
		}
		if !present {
			return Claims{}, &ChainMismatchError{ApplicationID: expected}
		}
	}

	return Claims{
		Sub:              sub,
		ZoneID:           zoneID,
		ClientID:         clientID,
		Sid:              sid,
		Scope:            scope,
		AgentSessionID:   agentSessionID,
		DelegationEdgeID: delegationEdgeID,
		SourceSessionID:  sourceSessionID,
		TargetSessionID:  targetSessionID,
		DelegationPath:   path,
		DelegationChain:  chain,
		GraphEpoch:       graphEpoch,
		HopCount:         hopCount,
	}, nil
}

// VerifyChainContains reports whether the claims include the given application
// either as an issuing party or in the delegation chain.
func VerifyChainContains(claims Claims, applicationID string) bool {
	if claims.ClientID == applicationID {
		return true
	}
	for _, hop := range claims.DelegationChain {
		if hop.ApplicationID == applicationID {
			return true
		}
	}
	return false
}

File: packages/identity/ts/src/verify.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Verifies a Caracal JWT against an issuer JWKS and enforces zone and scope claims.

import { jwtVerify } from 'jose'
import { hasScope } from '@caracalai/core'
import { getKeySet } from './jwks.js'
import type { ChainHop, Claims, JwtConfig } from './types.js'

export class TokenInvalidError extends Error {
  constructor(message = 'Token validation failed') {
    super(message)
    this.name = 'TokenInvalidError'
  }
}

export class ZoneInvalidError extends Error {
  constructor(message = 'Token zone validation failed') {
    super(message)
    this.name = 'ZoneInvalidError'
  }
}

export class ScopeInsufficientError extends Error {
  readonly missingScope: string
  constructor(missingScope: string) {
    super(`Missing scope: ${missingScope}`)
    this.name = 'ScopeInsufficientError'
    this.missingScope = missingScope
  }
}

export class AgentIdentityRequiredError extends Error {
  constructor(message = 'Agent identity required') {
    super(message)
    this.name = 'AgentIdentityRequiredError'
  }
}

export class DelegationRequiredError extends Error {
  constructor(message = 'Delegation required') {
    super(message)
    this.name = 'DelegationRequiredError'
  }
}

export class ChainMismatchError extends Error {
  readonly missingApplicationId: string
  constructor(missingApplicationId: string) {
    super(`Delegation chain missing application: ${missingApplicationId}`)
    this.name = 'ChainMismatchError'
    this.missingApplicationId = missingApplicationId
  }
}

function readChain(raw: unknown): ChainHop[] | undefined {
  if (!Array.isArray(raw)) return undefined
  const out: ChainHop[] = []
  for (const item of raw) {
    if (item && typeof item === 'object') {
      const r = item as Record<string, unknown>
      const applicationId = typeof r['app'] === 'string'
        ? r['app']
        : typeof r['application_id'] === 'string' ? r['application_id'] : ''
      if (!applicationId) continue
      out.push({
        applicationId,
        agentSessionId: typeof r['session'] === 'string'
          ? r['session']
          : typeof r['agent_session_id'] === 'string' ? r['agent_session_id'] as string : undefined,
        delegationEdgeId: typeof r['edge'] === 'string'
          ? r['edge']
          : typeof r['delegation_edge_id'] === 'string' ? r['delegation_edge_id'] as string : undefined,
      })
    }
  }
  return out.length === 0 ? undefined : out
}

export async function verify(token: string, config: JwtConfig): Promise<Claims> {
  let payload
  try {
    const keySet = await getKeySet(config.issuer)
    ;({ payload } = await jwtVerify(token, keySet, {
      issuer: config.issuer,
      audience: config.audience,
      algorithms: ['ES256'],
    }))
  } catch {
    throw new TokenInvalidError()
  }

  const scope = (payload['scope'] as string | undefined) ?? ''
  const zoneId = payload['zone_id']
  if (typeof zoneId !== 'string' || zoneId === '' || (config.zoneId && zoneId !== config.zoneId)) {
    throw new ZoneInvalidError()
  }
  for (const required of config.requiredScopes ?? []) {
    if (!hasScope(scope, required)) {
      throw new ScopeInsufficientError(required)
    }
  }

  const sid = typeof payload['sid'] === 'string' ? (payload['sid'] as string) : ''
  const clientId = typeof payload['client_id'] === 'string' ? (payload['client_id'] as string) : ''
  const agentSessionId = typeof payload['agent_session_id'] === 'string' ? (payload['agent_session_id'] as string) : undefined
  const delegationEdgeId = typeof payload['delegation_edge_id'] === 'string' ? (payload['delegation_edge_id'] as string) : undefined
  const sourceSessionId = typeof payload['source_session_id'] === 'string' ? (payload['source_session_id'] as string) : undefined
  const targetSessionId = typeof payload['target_session_id'] === 'string' ? (payload['target_session_id'] as string) : undefined
  const delegationPath = Array.isArray(payload['delegation_path'])
    ? (payload['delegation_path'] as unknown[]).filter((v) => typeof v === 'string') as string[]
    : undefined
  const delegationChain = readChain(payload['delegation_chain'])
  const graphEpoch = typeof payload['delegation_graph_epoch'] === 'number'
    ? payload['delegation_graph_epoch'] as number
    : (typeof payload['graph_epoch'] === 'number' ? payload['graph_epoch'] as number : undefined)
  const hopCount = typeof payload['hop_count'] === 'number' ? payload['hop_count'] as number : undefined

  if (config.requireAgent && !agentSessionId) {
    throw new AgentIdentityRequiredError()
  }
  if (config.requireDelegation && !delegationEdgeId) {
    throw new DelegationRequiredError()
  }
  for (const expected of config.requireChainContains ?? []) {
    const present = delegationChain?.some((h) => h.applicationId === expected)
    if (!present) throw new ChainMismatchError(expected)
  }

  return {
    sub: typeof payload.sub === 'string' ? payload.sub : '',
    zoneId,
    clientId,
    sid,
    scope,
    agentSessionId,
    delegationEdgeId,
    sourceSessionId,
    targetSessionId,
    delegationPath,
    delegationChain,
    graphEpoch,
    hopCount,
  }
}

export function verifyChainContains(claims: Claims, applicationId: string): boolean {
  if (claims.delegationChain?.some((h) => h.applicationId === applicationId)) return true
  if (claims.clientId === applicationId) return true
  return false
}
